Gravity compensated position control added

naveenoid opened the pull request:
Test incorporated inside kuka IIWA 

+@liangfok feature
+@sammy-tri platform

BTW : @hongkai-dai take a look if you have sometime.

Some items may change a bit depending on https://github.com/RobotLocomotion/drake/pull/3069
rebase will take care of it.

<!-- Reviewable:start -->
---
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/robotlocomotion/drake/3105)
<!-- Reviewable:end -->

naveenoid commented:
Temporarily marking do-not-review..I forgot to comment on some stuff.

sammy-tri commented:
Reviewed 9 of 9 files at r1.
Review status: all files reviewed at latest revision, 18 unresolved discussions, some commit checks failed.

---

*[drake/examples/kuka_iiwa_arm/iiwa_simulation.cc, line 4 \[r1\]](https://reviewable.io:443/reviews/robotlocomotion/drake/3105#-KOpl6_AftODqt2Zptil:-KOpl6_AftODqt2Zptim:b-618xyo) ([raw file](https://github.com/robotlocomotion/drake/blob/4c45ef209290d3cdf017ebc7105da7a65a7061ec/drake/examples/kuka_iiwa_arm/iiwa_simulation.cc#L4)):*
> ```cpp
> 
> #include "drake/common/drake_path.h"
> #include "drake/systems/LCMSystem.h"
> ```

This header doesn't appear to be used in this file?  (and probably wasn't before)

---

*[drake/examples/kuka_iiwa_arm/iiwa_simulation.h, line 4 \[r1\]](https://reviewable.io:443/reviews/robotlocomotion/drake/3105#-KOpknlfIeO5OAF56zK3:-KOpknlfIeO5OAF56zK4:bspng06) ([raw file](https://github.com/robotlocomotion/drake/blob/4c45ef209290d3cdf017ebc7105da7a65a7061ec/drake/examples/kuka_iiwa_arm/iiwa_simulation.h#L4)):*
> ```objc
> 
> #include <memory>
> #include <string>
> ```

This include doesn't seem to be used in this file?

---

*[drake/examples/kuka_iiwa_arm/iiwa_simulation.h, line 8 \[r1\]](https://reviewable.io:443/reviews/robotlocomotion/drake/3105#-KOpkRTG1NROSEyBKaCC:-KOpkRTG1NROSEyBKaCD:b-yiyf5j) ([raw file](https://github.com/robotlocomotion/drake/blob/4c45ef209290d3cdf017ebc7105da7a65a7061ec/drake/examples/kuka_iiwa_arm/iiwa_simulation.h#L8)):*
> ```objc
> #include <Eigen/Geometry>
> 
> #include "robot_state_tap.h"
> ```

This include doesn't appear to be used in this file?

---

*[drake/examples/kuka_iiwa_arm/iiwa_status.h, line 7 \[r1\]](https://reviewable.io:443/reviews/robotlocomotion/drake/3105#-KOpdbkkW29ZvNZTubwG:-KOpdbkkW29ZvNZTubwH:b4b2cxl) ([raw file](https://github.com/robotlocomotion/drake/blob/4c45ef209290d3cdf017ebc7105da7a65a7061ec/drake/examples/kuka_iiwa_arm/iiwa_status.h#L7)):*
> ```objc
> #include <Eigen/Core>
> 
> #include "drake/systems/LCMSystem.h"
> ```

I can't see where this header is used in this file?

---

*[drake/examples/kuka_iiwa_arm/test/kuka_iiwa_arm_test.cc, line 6 \[r1\]](https://reviewable.io:443/reviews/robotlocomotion/drake/3105#-KOpmITSTFiWJiWGveSi:-KOpmITSTFiWJiWGveSj:bfl9zbd) ([raw file](https://github.com/robotlocomotion/drake/blob/61715d69bb13d9d494b7f40530711deab02b3168/drake/examples/kuka_iiwa_arm/test/kuka_iiwa_arm_test.cc#L6)):*
> ```cpp
> 
> #include "drake/examples/kuka_iiwa_arm/iiwa_simulation.h"
> #include "drake/examples/kuka_iiwa_arm/robot_state_tap.h"
> ```

This include is used in this file.

---

*[drake/examples/kuka_iiwa_arm/test/kuka_iiwa_arm_test.cc, line 8 \[r1\]](https://reviewable.io:443/reviews/robotlocomotion/drake/3105#-KOpmM0Fpo4P3eIjy-5f:-KOpmM0G2UOhgXj4e6zH:bfl9zbd) ([raw file](https://github.com/robotlocomotion/drake/blob/61715d69bb13d9d494b7f40530711deab02b3168/drake/examples/kuka_iiwa_arm/test/kuka_iiwa_arm_test.cc#L8)):*
> ```cpp
> #include "drake/examples/kuka_iiwa_arm/robot_state_tap.h"
> #include "drake/common/drake_path.h"
> #include "drake/systems/cascade_system.h"
> ```

This include is used in this file.

---

*[drake/examples/kuka_iiwa_arm/test/kuka_iiwa_arm_test.cc, line 11 \[r1\]](https://reviewable.io:443/reviews/robotlocomotion/drake/3105#-KOpmSdmbqOfxJGyrDSF:-KOpmSdmbqOfxJGyrDSG:bfl9zbd) ([raw file](https://github.com/robotlocomotion/drake/blob/61715d69bb13d9d494b7f40530711deab02b3168/drake/examples/kuka_iiwa_arm/test/kuka_iiwa_arm_test.cc#L11)):*
> ```cpp
> #include "drake/systems/LCMSystem.h"
> #include "drake/systems/plants/BotVisualizer.h"
> #include "drake/systems/plants/RigidBodySystem.h"
> ```

This include is used in this file.

---

*[drake/examples/kuka_iiwa_arm/test/kuka_iiwa_arm_test.cc, line 12 \[r1\]](https://reviewable.io:443/reviews/robotlocomotion/drake/3105#-KOpmX3v4xZyQIXYkziW:-KOpmX3v4xZyQIXYkziX:bfl9zbd) ([raw file](https://github.com/robotlocomotion/drake/blob/61715d69bb13d9d494b7f40530711deab02b3168/drake/examples/kuka_iiwa_arm/test/kuka_iiwa_arm_test.cc#L12)):*
> ```cpp
> #include "drake/systems/plants/BotVisualizer.h"
> #include "drake/systems/plants/RigidBodySystem.h"
> #include "drake/systems/Simulation.h"
> ```

This include is used in this file.

---

*[drake/examples/kuka_iiwa_arm/test/kuka_iiwa_gravity_compensation_test.cc, line 16 \[r1\]](https://reviewable.io:443/reviews/robotlocomotion/drake/3105#-KOpn2SoVTqfwf8KLIH7:-KOpn2SoVTqfwf8KLIH8:b4yw4k3) ([raw file](https://github.com/robotlocomotion/drake/blob/4c45ef209290d3cdf017ebc7105da7a65a7061ec/drake/examples/kuka_iiwa_arm/test/kuka_iiwa_gravity_compensation_test.cc#L16)):*
> ```cpp
> using drake::GravityCompensatedPDPositionControlSystem;
> using Eigen::MatrixXd;
> using drake::RobotStateTap;
> ```

https://google.github.io/styleguide/cppguide.html#Names_and_Order_of_Includes -- "You should include all the headers that define the symbols you rely upon"  (not just this line, other parts of this file too)

---

*[drake/examples/kuka_iiwa_arm/test/kuka_iiwa_pd_control_test.cc, line 20 \[r1\]](https://reviewable.io:443/reviews/robotlocomotion/drake/3105#-KOpn_L4VXNLSCctyinE:-KOpn_L4VXNLSCctyinF:b827jjf) ([raw file](https://github.com/robotlocomotion/drake/blob/4c45ef209290d3cdf017ebc7105da7a65a7061ec/drake/examples/kuka_iiwa_arm/test/kuka_iiwa_pd_control_test.cc#L20)):*
> ```cpp
> using drake::PDControlSystem;
> using drake::RigidBodySystem;
> using drake::RobotStateTap;
> ```

Include what you use (as above)

---

*[drake/systems/gravity_compensation_control_system.h, line 7 \[r1\]](https://reviewable.io:443/reviews/robotlocomotion/drake/3105#-KOpe0owcROaMrudLdIN:-KOpe0oxbmntjRqHWZ6G:bruc22o) ([raw file](https://github.com/robotlocomotion/drake/blob/4c45ef209290d3cdf017ebc7105da7a65a7061ec/drake/systems/gravity_compensation_control_system.h#L7)):*
> ```objc
> #include "drake/common/drake_assert.h"
> #include "drake/systems/System.h"
> #include "drake/systems/plants/RigidBodySystem.h"
> ```

https://google.github.io/styleguide/cppguide.html#Names_and_Order_of_Includes "Within each section the includes should be ordered alphabetically."

---

*[drake/systems/gravity_compensation_control_system.h, line 11 \[r1\]](https://reviewable.io:443/reviews/robotlocomotion/drake/3105#-KOpeIH_V75fKHm192qt:-KOpeIH_V75fKHm192qu:b-k2ugv5) ([raw file](https://github.com/robotlocomotion/drake/blob/4c45ef209290d3cdf017ebc7105da7a65a7061ec/drake/systems/gravity_compensation_control_system.h#L11)):*
> ```objc
> #include "drake/systems/plants/KinematicsCache.h"
> 
> namespace drake {
> ```

BTW a couple of files in this directory are also in the "systems" namespace, but I don't think most are.  I'm not confident that I know the right thing to recommend, so I don't think only using the "drake" namespace is a problem here.

---

*[drake/systems/gravity_compensation_control_system.h, line 13 \[r1\]](https://reviewable.io:443/reviews/robotlocomotion/drake/3105#-KOpg1XNnO2oBWbXb3P0:-KOpg1XNnO2oBWbXb3P1:b-938t2z) ([raw file](https://github.com/robotlocomotion/drake/blob/4c45ef209290d3cdf017ebc7105da7a65a7061ec/drake/systems/gravity_compensation_control_system.h#L13)):*
> ```objc
> namespace drake {
> 
> /** GravityCompensatedPDPositionControlSystem<System>
> ```

This text on the first line after the ** is generated as the brief in doxygen.  We have JAVADOC_AUTOBRIEF enabled as discussed in http://www.stack.nl/~dimitri/doxygen/manual/docblocks.html

---

*[drake/systems/gravity_compensation_control_system.h, line 22 \[r1\]](https://reviewable.io:443/reviews/robotlocomotion/drake/3105#-KOph5vIs6_ZVJG0Z-jI:-KOph5vIs6_ZVJG0Z-jJ:b-xzfi0z) ([raw file](https://github.com/robotlocomotion/drake/blob/4c45ef209290d3cdf017ebc7105da7a65a7061ec/drake/systems/gravity_compensation_control_system.h#L22)):*
> ```objc
>                          |    [ G(x) ]<----------|
>  *                       |                       |
>  *                       ----[ Kp, Kd ]<----------
> ```

This ascii art renders as a single line in doxygen:

Implemented concepts:
    System Concept x_d —>[ Kp, Kd ]–>(+)--—>(+)-—>[ sys ]-—> yout | - | + | | [ G(x) ]<-------—| | | -—[ Kp, Kd ]<-------— 

---

*[drake/systems/gravity_compensation_control_system.h, line 26 \[r1\]](https://reviewable.io:443/reviews/robotlocomotion/drake/3105#-KOpgDgsBfIh5pZQL1vS:-KOpgDgsBfIh5pZQL1vT:b-t0beqq) ([raw file](https://github.com/robotlocomotion/drake/blob/4c45ef209290d3cdf017ebc7105da7a65a7061ec/drake/systems/gravity_compensation_control_system.h#L26)):*
> ```objc
>  */
> template <class System>
> class GravityCompensatedPDPositionControlSystem {
> ```

The name of the class differs from the name of the file, which seems confusing (gravity_compensat*ion* vs GravityCompensat*ed*Pd)

---

*[drake/systems/gravity_compensation_control_system.h, line 35 \[r1\]](https://reviewable.io:443/reviews/robotlocomotion/drake/3105#-KOpiLpSMIL8Dg3W1YOO:-KOpiLpSMIL8Dg3W1YOP:b-5qihq5) ([raw file](https://github.com/robotlocomotion/drake/blob/4c45ef209290d3cdf017ebc7105da7a65a7061ec/drake/systems/gravity_compensation_control_system.h#L35)):*
> ```objc
>   using OutputVector = typename System::template OutputVector<ScalarType>;
>   typedef std::shared_ptr<RigidBodySystem> RigidBodySystemPtr;
>   typedef std::shared_ptr<RigidBodyTree> RigidBodyTreePtr;
> ```

FYI I found this a little confusing since I didn't immediately recognize that it was a shared_ptr where the member fields were declared, but I'm not sure that rises to the level of a blocking defect.

---

*[drake/systems/gravity_compensation_control_system.h, line 39 \[r1\]](https://reviewable.io:443/reviews/robotlocomotion/drake/3105#-KOphSkjEn8nzncw0rIB:-KOphSkjEn8nzncw0rIC:b-ygplze) ([raw file](https://github.com/robotlocomotion/drake/blob/4c45ef209290d3cdf017ebc7105da7a65a7061ec/drake/systems/gravity_compensation_control_system.h#L39)):*
> ```objc
>   template <typename DerivedA, typename DerivedB>
>   GravityCompensatedPDPositionControlSystem(const RigidBodySystemPtr& sys,
>                                             const Eigen::MatrixBase<DerivedA>& Kp,
> ```

I'm not going to point out all of the cpplint errors in this file.  Run common/test/cpplint_wrapper.py

---

*[drake/systems/gravity_compensation_control_system.h, line 59 \[r1\]](https://reviewable.io:443/reviews/robotlocomotion/drake/3105#-KOpjGRj9eHLUjA3a8jm:-KOpjGRj9eHLUjA3a8jn:b5k1i8e) ([raw file](https://github.com/robotlocomotion/drake/blob/4c45ef209290d3cdf017ebc7105da7a65a7061ec/drake/systems/gravity_compensation_control_system.h#L59)):*
> ```objc
> 
>    size_t num_DoF = Kp_.cols();
>    KinematicsCache<double> cache_ = sys_tree_->doKinematics(toEigen(x).head(num_DoF), toEigen(x).tail(num_DoF));
> ```

Are we certain that the number of positions (q) plus the number of velocities (v) is equal to Kp_.cols() * 2?  or does Kd_.cols() need to be involed?  (it might be that they're always the same anyway)

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/3105)*
<!-- Sent from Reviewable.io -->

liangfok commented:
Review status: 8 of 9 files reviewed at latest revision, 18 unresolved discussions, some commit checks failed.

---

*[drake/systems/gravity_compensation_control_system.h, line 22 \[r1\]](https://reviewable.io:443/reviews/robotlocomotion/drake/3105#-KOph5vIs6_ZVJG0Z-jI:-KOqIfe35ezkR-7oi2rj:b-7tjyfu) ([raw file](https://github.com/robotlocomotion/drake/blob/4c45ef209290d3cdf017ebc7105da7a65a7061ec/drake/systems/gravity_compensation_control_system.h#L22)):*
<details><summary><i>Previously, sammy-tri wrote…</i></summary>
> This ascii art renders as a single line in doxygen:
> 
> Implemented concepts:
>     System Concept x_d —>[ Kp, Kd ]–>(+)--—>(+)-—>[ sys ]-—> yout | - | + | | [ G(x) ]<-------—| | | -—[ Kp, Kd ]<-------— 
</details>
To render correctly, surround it in:

```
<pre>

...

</pre>
```

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/3105)*
<!-- Sent from Reviewable.io -->

hongkai-dai commented:
Reviewed 8 of 9 files at r1, 1 of 1 files at r2.
Review status: all files reviewed at latest revision, 19 unresolved discussions, some commit checks failed.

---

*[drake/systems/gravity_compensation_control_system.h, line 65 \[r2\]](https://reviewable.io:443/reviews/robotlocomotion/drake/3105#-KOqlkSQ0Pu7oQ-KgU3h:-KOqlkSQ0Pu7oQ-KgU3i:b-cnd7lz) ([raw file](https://github.com/robotlocomotion/drake/blob/9de30dde56e1e4b70db8bf7f76196ce3e86ed350/drake/systems/gravity_compensation_control_system.h#L65)):*
> ```objc
>     Eigen::VectorXd vd(num_DoF);
>     vd.setZero();
>     auto G = sys_tree_->inverseDynamics(cache_, f_ext, vd, false);
> ```

Is this a temporary plan to use `inverseDynamics` to compute the gravitational vector? Some computation in inverseDynamics function is redundant, do we plan to create a separate function `generalizedGravity` to compute the gravitational vector in the future?

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/3105)*
<!-- Sent from Reviewable.io -->

liangfok commented:
Review status: all files reviewed at latest revision, 19 unresolved discussions, some commit checks failed.

---

*[drake/systems/gravity_compensation_control_system.h, line 65 \[r2\]](https://reviewable.io:443/reviews/robotlocomotion/drake/3105#-KOqlkSQ0Pu7oQ-KgU3h:-KOqnIa1vvXxI8CPaxhq:b8zhnfj) ([raw file](https://github.com/robotlocomotion/drake/blob/9de30dde56e1e4b70db8bf7f76196ce3e86ed350/drake/systems/gravity_compensation_control_system.h#L65)):*
<details><summary><i>Previously, hongkai-dai (Hongkai Dai) wrote…</i></summary>
> Is this a temporary plan to use `inverseDynamics` to compute the gravitational vector? Some computation in inverseDynamics function is redundant, do we plan to create a separate function `generalizedGravity` to compute the gravitational vector in the future?
</details>
Good point. I just created: https://github.com/RobotLocomotion/drake/issues/3114

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/3105)*
<!-- Sent from Reviewable.io -->

naveenoid commented:
removed do-not-review as the review is anyways underway..Will putt in some work in the next commit to address your concerns.

sammy-tri commented:
Sorry, I somehow missed the do-not-review tag the first time (though it was obviously there).  I'll certainly attempt to pay more attention in the future.

---

Reviewed 3 of 3 files at r3.
Review status: all files reviewed at latest revision, 19 unresolved discussions, some commit checks failed.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/3105#-:-KOugO0iv-fpUhx4LcXZ:b-c45o)*
<!-- Sent from Reviewable.io -->

liangfok commented:
I left a few comments, and will do a full feature review pass once this PR is rebased with the latest on master.

---

Reviewed 5 of 9 files at r1, 1 of 1 files at r2.
Review status: all files reviewed at latest revision, 24 unresolved discussions, some commit checks failed.

---

*[drake/examples/kuka_iiwa_arm/iiwa_simulation.h, line 37 \[r3\]](https://reviewable.io:443/reviews/robotlocomotion/drake/3105#-KOw85jXTUMQtjIY1GlD:-KOw85jXTUMQtjIY1GlE:b-vli47) ([raw file](https://github.com/robotlocomotion/drake/blob/225ed58e0893aaa6f496eaf8f04a32a86a91f99d/drake/examples/kuka_iiwa_arm/iiwa_simulation.h#L37)):*
> ```objc
> drake::SimulationOptions SetupSimulation(double initial_step_size = 0.002);
> 
> /*
> ```

Since this is a Doxygen comment block, this should be `/**`.

---

*[drake/examples/kuka_iiwa_arm/iiwa_simulation.h, line 38 \[r3\]](https://reviewable.io:443/reviews/robotlocomotion/drake/3105#-KOw8FKdK0NMAHWHIAt6:-KOw8FKdK0NMAHWHIAt7:by6puic) ([raw file](https://github.com/robotlocomotion/drake/blob/225ed58e0893aaa6f496eaf8f04a32a86a91f99d/drake/examples/kuka_iiwa_arm/iiwa_simulation.h#L38)):*
> ```objc
> 
> /*
>  * Check for joint position and joint velocity limit violations.
> ```

"Check" should be "Checks"

Consider deleting the second "joint" for conciseness without sacrificing understandnig.

---

*[drake/examples/kuka_iiwa_arm/iiwa_status.h, line 90 \[r2\]](https://reviewable.io:443/reviews/robotlocomotion/drake/3105#-KOw8bxW4XmBIRMO5KyB:-KOw8bxW4XmBIRMO5KyC:bmt6ecf) ([raw file](https://github.com/robotlocomotion/drake/blob/9de30dde56e1e4b70db8bf7f76196ce3e86ed350/drake/examples/kuka_iiwa_arm/iiwa_status.h#L90)):*
> ```objc
>   return true;
> }
> }  // namespace kuka_iiwa_arm
> ```

I think there should be one blank line above `}  // namespace kuka_iiwa_arm`.

---

*[drake/systems/gravity_compensation_control_system.h, line 112 \[r2\]](https://reviewable.io:443/reviews/robotlocomotion/drake/3105#-KOw9Wx_CifB5a8ky6o6:-KOw9Wx_CifB5a8ky6o7:bki4f1h) ([raw file](https://github.com/robotlocomotion/drake/blob/9de30dde56e1e4b70db8bf7f76196ce3e86ed350/drake/systems/gravity_compensation_control_system.h#L112)):*
> ```objc
>   RigidBodyTreePtr sys_tree_;
>   Eigen::MatrixXd Kp_, Kd_;
> ```

No need for blank line here.

---

*[drake/systems/gravity_compensation_control_system.h, line 115 \[r2\]](https://reviewable.io:443/reviews/robotlocomotion/drake/3105#-KOw9TOFpCWRcLwGNqU5:-KOw9TOFpCWRcLwGNqU6:bu6yona) ([raw file](https://github.com/robotlocomotion/drake/blob/9de30dde56e1e4b70db8bf7f76196ce3e86ed350/drake/systems/gravity_compensation_control_system.h#L115)):*
> ```objc
> };
> ```

Too many blank lines here (should be one).

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/3105#-:-KOw94GDPOOgJt0ZRy3M:bvhjopp)*
<!-- Sent from Reviewable.io -->

naveenoid commented:
Review status: 2 of 10 files reviewed at latest revision, 24 unresolved discussions.

---

*[drake/examples/kuka_iiwa_arm/iiwa_simulation.cc, line 4 \[r1\]](https://reviewable.io:443/reviews/robotlocomotion/drake/3105#-KOpl6_AftODqt2Zptil:-KPIkT46j42caoFSgFvU:b-896fix) ([raw file](https://github.com/robotlocomotion/drake/blob/4c45ef209290d3cdf017ebc7105da7a65a7061ec/drake/examples/kuka_iiwa_arm/iiwa_simulation.cc#L4)):*
<details><summary><i>Previously, sammy-tri wrote…</i></summary>
> This header doesn't appear to be used in this file?  (and probably wasn't before)
</details>
Done.

---

*[drake/examples/kuka_iiwa_arm/iiwa_simulation.h, line 4 \[r1\]](https://reviewable.io:443/reviews/robotlocomotion/drake/3105#-KOpknlfIeO5OAF56zK3:-KPIkXR9-NnS0dbJ55FV:b-896fix) ([raw file](https://github.com/robotlocomotion/drake/blob/4c45ef209290d3cdf017ebc7105da7a65a7061ec/drake/examples/kuka_iiwa_arm/iiwa_simulation.h#L4)):*
<details><summary><i>Previously, sammy-tri wrote…</i></summary>
> This include doesn't seem to be used in this file?
</details>
Done.

---

*[drake/examples/kuka_iiwa_arm/iiwa_simulation.h, line 8 \[r1\]](https://reviewable.io:443/reviews/robotlocomotion/drake/3105#-KOpkRTG1NROSEyBKaCC:-KPIk_i4_gbYsa8UCrAT:b-896fix) ([raw file](https://github.com/robotlocomotion/drake/blob/4c45ef209290d3cdf017ebc7105da7a65a7061ec/drake/examples/kuka_iiwa_arm/iiwa_simulation.h#L8)):*
<details><summary><i>Previously, sammy-tri wrote…</i></summary>
> This include doesn't appear to be used in this file?
</details>
Done.

---

*[drake/examples/kuka_iiwa_arm/iiwa_simulation.h, line 37 \[r3\]](https://reviewable.io:443/reviews/robotlocomotion/drake/3105#-KOw85jXTUMQtjIY1GlD:-KPIkfC3_R3nshIR-dIU:b-896fix) ([raw file](https://github.com/robotlocomotion/drake/blob/225ed58e0893aaa6f496eaf8f04a32a86a91f99d/drake/examples/kuka_iiwa_arm/iiwa_simulation.h#L37)):*
<details><summary><i>Previously, liangfok (Chien-Liang Fok) wrote…</i></summary>
> Since this is a Doxygen comment block, this should be `/**`.
</details>
Done.

---

*[drake/examples/kuka_iiwa_arm/iiwa_simulation.h, line 38 \[r3\]](https://reviewable.io:443/reviews/robotlocomotion/drake/3105#-KOw8FKdK0NMAHWHIAt6:-KPIkjJZqeztxjyE4SqY:b-896fix) ([raw file](https://github.com/robotlocomotion/drake/blob/225ed58e0893aaa6f496eaf8f04a32a86a91f99d/drake/examples/kuka_iiwa_arm/iiwa_simulation.h#L38)):*
<details><summary><i>Previously, liangfok (Chien-Liang Fok) wrote…</i></summary>
> "Check" should be "Checks"
> 
> Consider deleting the second "joint" for conciseness without sacrificing understandnig.
</details>
Done.

---

*[drake/examples/kuka_iiwa_arm/iiwa_status.h, line 7 \[r1\]](https://reviewable.io:443/reviews/robotlocomotion/drake/3105#-KOpdbkkW29ZvNZTubwG:-KPIYFVdJIILUzDbbiHV:bss4a2b) ([raw file](https://github.com/robotlocomotion/drake/blob/4c45ef209290d3cdf017ebc7105da7a65a7061ec/drake/examples/kuka_iiwa_arm/iiwa_status.h#L7)):*
<details><summary><i>Previously, sammy-tri wrote…</i></summary>
> I can't see where this header is used in this file?
</details>
Removed.

---

*[drake/examples/kuka_iiwa_arm/iiwa_status.h, line 90 \[r2\]](https://reviewable.io:443/reviews/robotlocomotion/drake/3105#-KOw8bxW4XmBIRMO5KyB:-KPIWqxZDbTHZcDtbhAX:b-896fix) ([raw file](https://github.com/robotlocomotion/drake/blob/9de30dde56e1e4b70db8bf7f76196ce3e86ed350/drake/examples/kuka_iiwa_arm/iiwa_status.h#L90)):*
<details><summary><i>Previously, liangfok (Chien-Liang Fok) wrote…</i></summary>
> I think there should be one blank line above `}  // namespace kuka_iiwa_arm`.
</details>
Done.

---

*[drake/examples/kuka_iiwa_arm/test/kuka_iiwa_arm_test.cc, line 11 \[r1\]](https://reviewable.io:443/reviews/robotlocomotion/drake/3105#-KOpmSdmbqOfxJGyrDSF:-KPIX2p63mPuU6uic0-z:bm8pnbp) ([raw file](https://github.com/robotlocomotion/drake/blob/61715d69bb13d9d494b7f40530711deab02b3168/drake/examples/kuka_iiwa_arm/test/kuka_iiwa_arm_test.cc#L11)):*
<details><summary><i>Previously, sammy-tri wrote…</i></summary>
> This include is used in this file.
</details>
Im not fully sure what you mean..what change should I do?

---

*[drake/examples/kuka_iiwa_arm/test/kuka_iiwa_arm_test.cc, line 12 \[r1\]](https://reviewable.io:443/reviews/robotlocomotion/drake/3105#-KOpmX3v4xZyQIXYkziW:-KPIX9R3_zhR4LEcJzT6:bm8pnbp) ([raw file](https://github.com/robotlocomotion/drake/blob/61715d69bb13d9d494b7f40530711deab02b3168/drake/examples/kuka_iiwa_arm/test/kuka_iiwa_arm_test.cc#L12)):*
<details><summary><i>Previously, sammy-tri wrote…</i></summary>
> This include is used in this file.
</details>
Im not fully sure what you mean..what change should I do?

---

*[drake/examples/kuka_iiwa_arm/test/kuka_iiwa_gravity_compensation_test.cc, line 16 \[r1\]](https://reviewable.io:443/reviews/robotlocomotion/drake/3105#-KOpn2SoVTqfwf8KLIH7:-KPIXAab94N96c-8Rhy9:b-896fix) ([raw file](https://github.com/robotlocomotion/drake/blob/4c45ef209290d3cdf017ebc7105da7a65a7061ec/drake/examples/kuka_iiwa_arm/test/kuka_iiwa_gravity_compensation_test.cc#L16)):*
<details><summary><i>Previously, sammy-tri wrote…</i></summary>
> https://google.github.io/styleguide/cppguide.html#Names_and_Order_of_Includes -- "You should include all the headers that define the symbols you rely upon"  (not just this line, other parts of this file too)
</details>
Done.

---

*[drake/examples/kuka_iiwa_arm/test/kuka_iiwa_pd_control_test.cc, line 20 \[r1\]](https://reviewable.io:443/reviews/robotlocomotion/drake/3105#-KOpn_L4VXNLSCctyinE:-KPIlkt7aRMfxVbTDqDL:b-896fix) ([raw file](https://github.com/robotlocomotion/drake/blob/4c45ef209290d3cdf017ebc7105da7a65a7061ec/drake/examples/kuka_iiwa_arm/test/kuka_iiwa_pd_control_test.cc#L20)):*
<details><summary><i>Previously, sammy-tri wrote…</i></summary>
> Include what you use (as above)
</details>
Done.

---

*[drake/systems/gravity_compensation_control_system.h, line 7 \[r1\]](https://reviewable.io:443/reviews/robotlocomotion/drake/3105#-KOpe0owcROaMrudLdIN:-KPIXFE8vI7a91-13vIz:b-896fix) ([raw file](https://github.com/robotlocomotion/drake/blob/4c45ef209290d3cdf017ebc7105da7a65a7061ec/drake/systems/gravity_compensation_control_system.h#L7)):*
<details><summary><i>Previously, sammy-tri wrote…</i></summary>
> https://google.github.io/styleguide/cppguide.html#Names_and_Order_of_Includes "Within each section the includes should be ordered alphabetically."
</details>
Done.

---

*[drake/systems/gravity_compensation_control_system.h, line 13 \[r1\]](https://reviewable.io:443/reviews/robotlocomotion/drake/3105#-KOpg1XNnO2oBWbXb3P0:-KPIXJUZBl7CUuykowx_:b-896fix) ([raw file](https://github.com/robotlocomotion/drake/blob/4c45ef209290d3cdf017ebc7105da7a65a7061ec/drake/systems/gravity_compensation_control_system.h#L13)):*
<details><summary><i>Previously, sammy-tri wrote…</i></summary>
> This text on the first line after the ** is generated as the brief in doxygen.  We have JAVADOC_AUTOBRIEF enabled as discussed in http://www.stack.nl/~dimitri/doxygen/manual/docblocks.html
</details>
Done.

---

*[drake/systems/gravity_compensation_control_system.h, line 22 \[r1\]](https://reviewable.io:443/reviews/robotlocomotion/drake/3105#-KOph5vIs6_ZVJG0Z-jI:-KPIXKL2ySd7kpXSqKB3:bgm7u6m) ([raw file](https://github.com/robotlocomotion/drake/blob/4c45ef209290d3cdf017ebc7105da7a65a7061ec/drake/systems/gravity_compensation_control_system.h#L22)):*
<details><summary><i>Previously, liangfok (Chien-Liang Fok) wrote…</i></summary>
> To render correctly, surround it in:
> 
> ```
> <pre>
> 
> ...
> 
> </pre>
> ```
</details>
I will tentatively link to a block diagram pushed in drake/doc/images.. since its not in master yet, I will link to my forks dev branch..future PR to link to master will be an easy update.

---

*[drake/systems/gravity_compensation_control_system.h, line 26 \[r1\]](https://reviewable.io:443/reviews/robotlocomotion/drake/3105#-KOpgDgsBfIh5pZQL1vS:-KPIYJvXJ3pFxJUCIEf6:b-y12qi7) ([raw file](https://github.com/robotlocomotion/drake/blob/4c45ef209290d3cdf017ebc7105da7a65a7061ec/drake/systems/gravity_compensation_control_system.h#L26)):*
<details><summary><i>Previously, sammy-tri wrote…</i></summary>
> The name of the class differs from the name of the file, which seems confusing (gravity_compensat*ion* vs GravityCompensat*ed*Pd)
</details>
I intend to also add the Gravity Compensated torque controller to this header to keep them in the same place. Hence the header name. I will change it if you feel its a defect. Alternately I can add a TODO for the class that is to be added.

---

*[drake/systems/gravity_compensation_control_system.h, line 35 \[r1\]](https://reviewable.io:443/reviews/robotlocomotion/drake/3105#-KOpiLpSMIL8Dg3W1YOO:-KPIcQp941oEzKKASuSI:byq9z58) ([raw file](https://github.com/robotlocomotion/drake/blob/4c45ef209290d3cdf017ebc7105da7a65a7061ec/drake/systems/gravity_compensation_control_system.h#L35)):*
<details><summary><i>Previously, sammy-tri wrote…</i></summary>
> FYI I found this a little confusing since I didn't immediately recognize that it was a shared_ptr where the member fields were declared, but I'm not sure that rises to the level of a blocking defect.
</details>
Added a comment to clarify.

---

*[drake/systems/gravity_compensation_control_system.h, line 39 \[r1\]](https://reviewable.io:443/reviews/robotlocomotion/drake/3105#-KOphSkjEn8nzncw0rIB:-KPIhvRRKTbm3WtKP4EA:b-896fix) ([raw file](https://github.com/robotlocomotion/drake/blob/4c45ef209290d3cdf017ebc7105da7a65a7061ec/drake/systems/gravity_compensation_control_system.h#L39)):*
<details><summary><i>Previously, sammy-tri wrote…</i></summary>
> I'm not going to point out all of the cpplint errors in this file.  Run common/test/cpplint_wrapper.py
</details>
Done.

---

*[drake/systems/gravity_compensation_control_system.h, line 59 \[r1\]](https://reviewable.io:443/reviews/robotlocomotion/drake/3105#-KOpjGRj9eHLUjA3a8jm:-KPIiONop2W1sCCowk2C:b-hwxw8c) ([raw file](https://github.com/robotlocomotion/drake/blob/4c45ef209290d3cdf017ebc7105da7a65a7061ec/drake/systems/gravity_compensation_control_system.h#L59)):*
<details><summary><i>Previously, sammy-tri wrote…</i></summary>
> Are we certain that the number of positions (q) plus the number of velocities (v) is equal to Kp_.cols() * 2?  or does Kd_.cols() need to be involed?  (it might be that they're always the same anyway)
</details>
Yes it is always equal. 

---

*[drake/systems/gravity_compensation_control_system.h, line 65 \[r2\]](https://reviewable.io:443/reviews/robotlocomotion/drake/3105#-KOqlkSQ0Pu7oQ-KgU3h:-KPIiZxIZ-4MwUClH4AC:b-asfhbb) ([raw file](https://github.com/robotlocomotion/drake/blob/9de30dde56e1e4b70db8bf7f76196ce3e86ed350/drake/systems/gravity_compensation_control_system.h#L65)):*
<details><summary><i>Previously, liangfok (Chien-Liang Fok) wrote…</i></summary>
> Good point. I just created: https://github.com/RobotLocomotion/drake/issues/3114
</details>
I wont fix it here but I added a TODO 

---

*[drake/systems/gravity_compensation_control_system.h, line 112 \[r2\]](https://reviewable.io:443/reviews/robotlocomotion/drake/3105#-KOw9Wx_CifB5a8ky6o6:-KPIkIDcJ6r_riDSYF7R:b-896fix) ([raw file](https://github.com/robotlocomotion/drake/blob/9de30dde56e1e4b70db8bf7f76196ce3e86ed350/drake/systems/gravity_compensation_control_system.h#L112)):*
<details><summary><i>Previously, liangfok (Chien-Liang Fok) wrote…</i></summary>
> No need for blank line here.
</details>
Done.

---

*[drake/systems/gravity_compensation_control_system.h, line 115 \[r2\]](https://reviewable.io:443/reviews/robotlocomotion/drake/3105#-KOw9TOFpCWRcLwGNqU5:-KPIkIo1-5zrYWhUURvO:b-896fix) ([raw file](https://github.com/robotlocomotion/drake/blob/9de30dde56e1e4b70db8bf7f76196ce3e86ed350/drake/systems/gravity_compensation_control_system.h#L115)):*
<details><summary><i>Previously, liangfok (Chien-Liang Fok) wrote…</i></summary>
> Too many blank lines here (should be one).
</details>
Done.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/3105)*
<!-- Sent from Reviewable.io -->

naveenoid commented:
Ive rebased and addressed the issues..PTAL.

---

Review status: 2 of 10 files reviewed at latest revision, 24 unresolved discussions.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/3105#-:-KPImGVNh0vzzYEYk1Y6:btcgl77)*
<!-- Sent from Reviewable.io -->

liangfok commented:
Completed another pass. PTAL. Thanks!

---

Reviewed 7 of 8 files at r5.
Review status: 9 of 10 files reviewed at latest revision, 22 unresolved discussions.

---

*[drake/examples/kuka_iiwa_arm/test/kuka_iiwa_gravity_compensation_test.cc, line 27 \[r5\]](https://reviewable.io:443/reviews/robotlocomotion/drake/3105#-KPIrnaUq3iFq4rSjF6G:-KPIrnaUq3iFq4rSjF6H:bkr9qvs) ([raw file](https://github.com/robotlocomotion/drake/blob/283c9a1b03c8805e3f06c71ff48aaa7d93996084/drake/examples/kuka_iiwa_arm/test/kuka_iiwa_gravity_compensation_test.cc#L27)):*
> ```cpp
> namespace {
> 
> // Test to verify behaviour of the KUKA IIWA Arm under a gravity
> ```

UK spelling of "behavior".

---

*[drake/examples/kuka_iiwa_arm/test/kuka_iiwa_gravity_compensation_test.cc, line 53 \[r5\]](https://reviewable.io:443/reviews/robotlocomotion/drake/3105#-KPIs17Mu3jzxAcx6Qgi:-KPIs17Mu3jzxAcx6Qgj:brs1fe9) ([raw file](https://github.com/robotlocomotion/drake/blob/283c9a1b03c8805e3f06c71ff48aaa7d93996084/drake/examples/kuka_iiwa_arm/test/kuka_iiwa_gravity_compensation_test.cc#L53)):*
> ```cpp
>   int num_dof = iiwa_tree->number_of_positions();
> 
>   // Smaller gains intentionally used for demo.
> ```

BTW are these gains really unit less? Don't we need to normalize them against the maximum error for them to be unit less? Since we are not doing that, what are the units of these gains?

---

*[drake/systems/gravity_compensation_control_system.h, line 22 \[r1\]](https://reviewable.io:443/reviews/robotlocomotion/drake/3105#-KOph5vIs6_ZVJG0Z-jI:-KPItD5Gq8FITiDJFqi9:b-ezptnd) ([raw file](https://github.com/robotlocomotion/drake/blob/4c45ef209290d3cdf017ebc7105da7a65a7061ec/drake/systems/gravity_compensation_control_system.h#L22)):*
<details><summary><i>Previously, naveenoid (Naveen Kuppuswamy) wrote…</i></summary>
> I will tentatively link to a block diagram pushed in drake/doc/images.. since its not in master yet, I will link to my forks dev branch..future PR to link to master will be an easy update.
</details>
Wow. Instead of embedding the image data in this `.h` file, I recommend using a `<img src="example.jpg"/>`. It should be possible:

http://doxygen.10944.n7.nabble.com/How-to-create-image-link-in-doxygen-html-output-td1641.html

---

*[drake/systems/gravity_compensation_control_system.h, line 26 \[r1\]](https://reviewable.io:443/reviews/robotlocomotion/drake/3105#-KOpgDgsBfIh5pZQL1vS:-KPItXyC18xL-G8w1_W8:b7myil9) ([raw file](https://github.com/robotlocomotion/drake/blob/4c45ef209290d3cdf017ebc7105da7a65a7061ec/drake/systems/gravity_compensation_control_system.h#L26)):*
<details><summary><i>Previously, naveenoid (Naveen Kuppuswamy) wrote…</i></summary>
> I intend to also add the Gravity Compensated torque controller to this header to keep them in the same place. Hence the header name. I will change it if you feel its a defect. Alternately I can add a TODO for the class that is to be added.
</details>
I vote for placing less code in a single file and having more files, ideally with file names that match class names.

---

*[drake/systems/gravity_compensation_control_system.h, line 85 \[r5\]](https://reviewable.io:443/reviews/robotlocomotion/drake/3105#-KPItkACS2qEUNluPCHy:-KPItkACS2qEUNluPCHz:bwhz90f) ([raw file](https://github.com/robotlocomotion/drake/blob/283c9a1b03c8805e3f06c71ff48aaa7d93996084/drake/systems/gravity_compensation_control_system.h#L85)):*
> ```objc
>     vd.setZero();
> 
>     // The gravity torque are computed by calling inverse dynamics with
> ```

"torque" should be "torques". 

Also, this is assuming the robot has no prismatic joints. Consider "generalized effort values"?

---

*[drake/systems/gravity_compensation_control_system.h, line 88 \[r5\]](https://reviewable.io:443/reviews/robotlocomotion/drake/3105#-KPItowIrXjUPfPeP01-:-KPItowIrXjUPfPeP010:bldms2c) ([raw file](https://github.com/robotlocomotion/drake/blob/283c9a1b03c8805e3f06c71ff48aaa7d93996084/drake/systems/gravity_compensation_control_system.h#L88)):*
> ```objc
>     // 0 external forces, 0 velocities and 0 accelerations.
>     // TODO(naveenoid): Update to use simpler API once issue #3114 is
>     // resolved
> ```

Missing period at end of sentence.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/3105#-:-KPIu6Th-XrckbyyP-s9:b-aus8eb)*
<!-- Sent from Reviewable.io -->

sammy-tri commented:
Reviewed 7 of 8 files at r5.
Review status: 9 of 10 files reviewed at latest revision, 11 unresolved discussions.

---

*[drake/examples/kuka_iiwa_arm/test/kuka_iiwa_arm_test.cc, line 11 \[r1\]](https://reviewable.io:443/reviews/robotlocomotion/drake/3105#-KOpmSdmbqOfxJGyrDSF:-KPIx5_5dGKOp3APwYWk:b-q9nk75) ([raw file](https://github.com/robotlocomotion/drake/blob/61715d69bb13d9d494b7f40530711deab02b3168/drake/examples/kuka_iiwa_arm/test/kuka_iiwa_arm_test.cc#L11)):*
<details><summary><i>Previously, naveenoid (Naveen Kuppuswamy) wrote…</i></summary>
> Im not fully sure what you mean..what change should I do?
</details>
My issue was that it seemed strange to delete a #include line for a class which is actually used in that .cc file. (which RigidBodySystem is)  Also it  looks like it's back in the current version.

---

*[drake/examples/kuka_iiwa_arm/test/kuka_iiwa_arm_test.cc, line 12 \[r1\]](https://reviewable.io:443/reviews/robotlocomotion/drake/3105#-KOpmX3v4xZyQIXYkziW:-KPIxPSd5Kc3L1Zpg0Ew:bw2sgfy) ([raw file](https://github.com/robotlocomotion/drake/blob/61715d69bb13d9d494b7f40530711deab02b3168/drake/examples/kuka_iiwa_arm/test/kuka_iiwa_arm_test.cc#L12)):*
<details><summary><i>Previously, naveenoid (Naveen Kuppuswamy) wrote…</i></summary>
> Im not fully sure what you mean..what change should I do?
</details>
the simulate() function is called later in the test, and is defined in Simulation.h.  It seemed odd to remove a header which is used later in the file.

---

*[drake/examples/kuka_iiwa_arm/test/kuka_iiwa_gravity_compensation_test.cc, line 27 \[r5\]](https://reviewable.io:443/reviews/robotlocomotion/drake/3105#-KPIrnaUq3iFq4rSjF6G:-KPIyRGrM27vJ670cEjY:bcs4e00) ([raw file](https://github.com/robotlocomotion/drake/blob/283c9a1b03c8805e3f06c71ff48aaa7d93996084/drake/examples/kuka_iiwa_arm/test/kuka_iiwa_gravity_compensation_test.cc#L27)):*
<details><summary><i>Previously, liangfok (Chien-Liang Fok) wrote…</i></summary>
> UK spelling of "behavior".
</details>
Is this a problem?  Generally speaking, I would be fine with US or UK spellings of most comments.

---

*[drake/systems/gravity_compensation_control_system.h, line 13 \[r1\]](https://reviewable.io:443/reviews/robotlocomotion/drake/3105#-KOpg1XNnO2oBWbXb3P0:-KPIyxCOrl4LUdVXTlaf:bv93pus) ([raw file](https://github.com/robotlocomotion/drake/blob/4c45ef209290d3cdf017ebc7105da7a65a7061ec/drake/systems/gravity_compensation_control_system.h#L13)):*
<details><summary><i>Previously, naveenoid (Naveen Kuppuswamy) wrote…</i></summary>
> Done.
</details>
Is the text "GravityCompensatedPDPositionControlSystem<System>" actually what you want to have appear for the brief description of the class?  I think that can be removed.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/3105)*
<!-- Sent from Reviewable.io -->

liangfok commented:
*[drake/examples/kuka_iiwa_arm/test/kuka_iiwa_gravity_compensation_test.cc, line 27 \[r5\]](https://reviewable.io:443/reviews/robotlocomotion/drake/3105#-KPIrnaUq3iFq4rSjF6G:-KPJ0KjbvSSTDTaQ-6EG:b-ypykte) ([raw file](https://github.com/robotlocomotion/drake/blob/283c9a1b03c8805e3f06c71ff48aaa7d93996084/drake/examples/kuka_iiwa_arm/test/kuka_iiwa_gravity_compensation_test.cc#L27)):*
<details><summary><i>Previously, sammy-tri wrote…</i></summary>
> Is this a problem?  Generally speaking, I would be fine with US or UK spellings of most comments.
</details>
I concede there's no rule specifying either way. I just assume consistently using US English is preferred. Sorry, I should have marked this as "BTW".

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/3105)*
<!-- Sent from Reviewable.io -->

naveenoid commented:
Review status: 9 of 10 files reviewed at latest revision, 9 unresolved discussions.

---

*[drake/examples/kuka_iiwa_arm/test/kuka_iiwa_arm_test.cc, line 11 \[r1\]](https://reviewable.io:443/reviews/robotlocomotion/drake/3105#-KOpmSdmbqOfxJGyrDSF:-KPJ-mn0Y4Jjp5yKRxnw:bxrb941) ([raw file](https://github.com/robotlocomotion/drake/blob/61715d69bb13d9d494b7f40530711deab02b3168/drake/examples/kuka_iiwa_arm/test/kuka_iiwa_arm_test.cc#L11)):*
<details><summary><i>Previously, sammy-tri wrote…</i></summary>
> My issue was that it seemed strange to delete a #include line for a class which is actually used in that .cc file. (which RigidBodySystem is)  Also it  looks like it's back in the current version.
</details>
Done. I understand that now. 

---

*[drake/examples/kuka_iiwa_arm/test/kuka_iiwa_arm_test.cc, line 12 \[r1\]](https://reviewable.io:443/reviews/robotlocomotion/drake/3105#-KOpmX3v4xZyQIXYkziW:-KPJ-qDMsa-6FDb9ZQIj:b-896fix) ([raw file](https://github.com/robotlocomotion/drake/blob/61715d69bb13d9d494b7f40530711deab02b3168/drake/examples/kuka_iiwa_arm/test/kuka_iiwa_arm_test.cc#L12)):*
<details><summary><i>Previously, sammy-tri wrote…</i></summary>
> the simulate() function is called later in the test, and is defined in Simulation.h.  It seemed odd to remove a header which is used later in the file.
</details>
Done.

---

*[drake/examples/kuka_iiwa_arm/test/kuka_iiwa_gravity_compensation_test.cc, line 27 \[r5\]](https://reviewable.io:443/reviews/robotlocomotion/drake/3105#-KPIrnaUq3iFq4rSjF6G:-KPJ0KMok3QRJfmMWVzr:b-a77p12) ([raw file](https://github.com/robotlocomotion/drake/blob/283c9a1b03c8805e3f06c71ff48aaa7d93996084/drake/examples/kuka_iiwa_arm/test/kuka_iiwa_gravity_compensation_test.cc#L27)):*
<details><summary><i>Previously, sammy-tri wrote…</i></summary>
> Is this a problem?  Generally speaking, I would be fine with US or UK spellings of most comments.
</details>
Done. I changed in any case. I need to configure the spell check on my IDE!

---

*[drake/examples/kuka_iiwa_arm/test/kuka_iiwa_gravity_compensation_test.cc, line 53 \[r5\]](https://reviewable.io:443/reviews/robotlocomotion/drake/3105#-KPIs17Mu3jzxAcx6Qgi:-KPIuhCmqe3te76TQMoI:b-ra0lbu) ([raw file](https://github.com/robotlocomotion/drake/blob/283c9a1b03c8805e3f06c71ff48aaa7d93996084/drake/examples/kuka_iiwa_arm/test/kuka_iiwa_gravity_compensation_test.cc#L53)):*
<details><summary><i>Previously, liangfok (Chien-Liang Fok) wrote…</i></summary>
> BTW are these gains really unit less? Don't we need to normalize them against the maximum error for them to be unit less? Since we are not doing that, what are the units of these gains?
</details>
In this case the radians are Kp -> Nm/rad and for Kd -> Nm/rad/sec. Shall I add it inside the same comment or besides the next two lines?

---

*[drake/systems/gravity_compensation_control_system.h, line 11 \[r1\]](https://reviewable.io:443/reviews/robotlocomotion/drake/3105#-KOpeIH_V75fKHm192qt:-KPJ-X5mdJf1IHoP2Ypk:b-fkw78m) ([raw file](https://github.com/robotlocomotion/drake/blob/4c45ef209290d3cdf017ebc7105da7a65a7061ec/drake/systems/gravity_compensation_control_system.h#L11)):*
<details><summary><i>Previously, sammy-tri wrote…</i></summary>
> BTW a couple of files in this directory are also in the "systems" namespace, but I don't think most are.  I'm not confident that I know the right thing to recommend, so I don't think only using the "drake" namespace is a problem here.
</details>
Should I address this somehow?

---

*[drake/systems/gravity_compensation_control_system.h, line 13 \[r1\]](https://reviewable.io:443/reviews/robotlocomotion/drake/3105#-KOpg1XNnO2oBWbXb3P0:-KPJ-BGt7B7eQ6GNwwqn:b-896fix) ([raw file](https://github.com/robotlocomotion/drake/blob/4c45ef209290d3cdf017ebc7105da7a65a7061ec/drake/systems/gravity_compensation_control_system.h#L13)):*
<details><summary><i>Previously, sammy-tri wrote…</i></summary>
> Is the text "GravityCompensatedPDPositionControlSystem<System>" actually what you want to have appear for the brief description of the class?  I think that can be removed.
</details>
Done.

---

*[drake/systems/gravity_compensation_control_system.h, line 22 \[r1\]](https://reviewable.io:443/reviews/robotlocomotion/drake/3105#-KOph5vIs6_ZVJG0Z-jI:-KPJ-EEDpAPWXLKPhWFG:bkq2eqy) ([raw file](https://github.com/robotlocomotion/drake/blob/4c45ef209290d3cdf017ebc7105da7a65a7061ec/drake/systems/gravity_compensation_control_system.h#L22)):*
<details><summary><i>Previously, liangfok (Chien-Liang Fok) wrote…</i></summary>
> Wow. Instead of embedding the image data in this `.h` file, I recommend using a `<img src="example.jpg"/>`. It should be possible:
> 
> http://doxygen.10944.n7.nabble.com/How-to-create-image-link-in-doxygen-html-output-td1641.html
</details>
Its frustrating..Although the file is pushed into my branch drake/doc/images, I cant link to it in that way. The obvious way seems to be to host the file somewhere and then link to it. Any suggestions on hosting service that is not github itself?

---

*[drake/systems/gravity_compensation_control_system.h, line 26 \[r1\]](https://reviewable.io:443/reviews/robotlocomotion/drake/3105#-KOpgDgsBfIh5pZQL1vS:-KPIvKQiGD7kr13QyOFj:b-fypwbj) ([raw file](https://github.com/robotlocomotion/drake/blob/4c45ef209290d3cdf017ebc7105da7a65a7061ec/drake/systems/gravity_compensation_control_system.h#L26)):*
<details><summary><i>Previously, liangfok (Chien-Liang Fok) wrote…</i></summary>
> I vote for placing less code in a single file and having more files, ideally with file names that match class names.
</details>
Done. Refactored to match class name.

---

*[drake/systems/gravity_compensation_control_system.h, line 85 \[r5\]](https://reviewable.io:443/reviews/robotlocomotion/drake/3105#-KPItkACS2qEUNluPCHy:-KPJ-SZHXlzx1r78X-ay:b-896fix) ([raw file](https://github.com/robotlocomotion/drake/blob/283c9a1b03c8805e3f06c71ff48aaa7d93996084/drake/systems/gravity_compensation_control_system.h#L85)):*
<details><summary><i>Previously, liangfok (Chien-Liang Fok) wrote…</i></summary>
> "torque" should be "torques". 
> 
> Also, this is assuming the robot has no prismatic joints. Consider "generalized effort values"?
</details>
Done.

---

*[drake/systems/gravity_compensation_control_system.h, line 88 \[r5\]](https://reviewable.io:443/reviews/robotlocomotion/drake/3105#-KPItowIrXjUPfPeP01-:-KPJ-UVoXOdxWQadLgFE:b-896fix) ([raw file](https://github.com/robotlocomotion/drake/blob/283c9a1b03c8805e3f06c71ff48aaa7d93996084/drake/systems/gravity_compensation_control_system.h#L88)):*
<details><summary><i>Previously, liangfok (Chien-Liang Fok) wrote…</i></summary>
> Missing period at end of sentence.
</details>
Done.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/3105)*
<!-- Sent from Reviewable.io -->

naveenoid commented:
Review status: 4 of 10 files reviewed at latest revision, 9 unresolved discussions.

---

*[drake/systems/gravity_compensated_pd_position_control_system.h, line 22 \[r1\]](https://reviewable.io:443/reviews/robotlocomotion/drake/3105#-KOph5vIs6_ZVJG0Z-jI:-KPJ1IuCLoQolhyzMCOb:brr249s) ([raw file](https://github.com/robotlocomotion/drake/blob/4c45ef209290d3cdf017ebc7105da7a65a7061ec/drake/systems/gravity_compensated_pd_position_control_system.h#L22)):*
<details><summary><i>Previously, naveenoid (Naveen Kuppuswamy) wrote…</i></summary>
> Its frustrating..Although the file is pushed into my branch drake/doc/images, I cant link to it in that way. The obvious way seems to be to host the file somewhere and then link to it. Any suggestions on hosting service that is not github itself?
</details>
FYI I linked to it but for some reason my local make documentation does not generate it with the image embedded..

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/3105)*
<!-- Sent from Reviewable.io -->

sammy-tri commented:
Reviewed 4 of 6 files at r6.
Review status: 7 of 10 files reviewed at latest revision, 3 unresolved discussions, some commit checks failed.

---

*[drake/systems/gravity_compensated_pd_position_control_system.h, line 11 \[r1\]](https://reviewable.io:443/reviews/robotlocomotion/drake/3105#-KOpeIH_V75fKHm192qt:-KPJECIDU2wRF8xV_CXM:bbn04hz) ([raw file](https://github.com/robotlocomotion/drake/blob/4c45ef209290d3cdf017ebc7105da7a65a7061ec/drake/systems/gravity_compensated_pd_position_control_system.h#L11)):*
<details><summary><i>Previously, naveenoid (Naveen Kuppuswamy) wrote…</i></summary>
> Should I address this somehow?
</details>
No, I was just whining, I think.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/3105)*
<!-- Sent from Reviewable.io -->

liangfok commented:
Just one comment about the image.

---

Reviewed 1 of 9 files at r5, 3 of 6 files at r6.
Review status: all files reviewed at latest revision, 3 unresolved discussions, some commit checks failed.

---

*[drake/examples/kuka_iiwa_arm/test/kuka_iiwa_gravity_compensation_test.cc, line 53 \[r5\]](https://reviewable.io:443/reviews/robotlocomotion/drake/3105#-KPIs17Mu3jzxAcx6Qgi:-KPJErweaNHHG1CvZujz:bix0rt6) ([raw file](https://github.com/robotlocomotion/drake/blob/283c9a1b03c8805e3f06c71ff48aaa7d93996084/drake/examples/kuka_iiwa_arm/test/kuka_iiwa_gravity_compensation_test.cc#L53)):*
<details><summary><i>Previously, naveenoid (Naveen Kuppuswamy) wrote…</i></summary>
> In this case the radians are Kp -> Nm/rad and for Kd -> Nm/rad/sec. Shall I add it inside the same comment or besides the next two lines?
</details>
Looks good the way you have it :smile: .

---

*[drake/systems/gravity_compensated_pd_position_control_system.h, line 22 \[r1\]](https://reviewable.io:443/reviews/robotlocomotion/drake/3105#-KOph5vIs6_ZVJG0Z-jI:-KPJC-92lKgqbPhc5X8Z:bxjsb5x) ([raw file](https://github.com/robotlocomotion/drake/blob/4c45ef209290d3cdf017ebc7105da7a65a7061ec/drake/systems/gravity_compensated_pd_position_control_system.h#L22)):*
<details><summary><i>Previously, naveenoid (Naveen Kuppuswamy) wrote…</i></summary>
> FYI I linked to it but for some reason my local make documentation does not generate it with the image embedded..
</details>
We should be able to anticipate the URL of a file once it's committed to master. Something like:

```
<img src="https://github.com/RobotLocomotion/drake/tree/master/drake/doc/images/foo.png" />
```

According to [this page](https://www.stack.nl/~dimitri/doxygen/manual/htmlcmds.html) the above can be directly included within a Doxygen comment block.

---

*[drake/systems/gravity_compensated_pd_position_control_system.h, line 26 \[r1\]](https://reviewable.io:443/reviews/robotlocomotion/drake/3105#-KOpgDgsBfIh5pZQL1vS:-KPJBw3H9V5ziuyedhev:b-ucfknh) ([raw file](https://github.com/robotlocomotion/drake/blob/4c45ef209290d3cdf017ebc7105da7a65a7061ec/drake/systems/gravity_compensated_pd_position_control_system.h#L26)):*
<details><summary><i>Previously, naveenoid (Naveen Kuppuswamy) wrote…</i></summary>
> Done. Refactored to match class name.
</details>
Thanks!

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/3105#-:-KPJEzFCiqqQ62xMWSOs:b-6bwt4x)*
<!-- Sent from Reviewable.io -->

liangfok commented:
The `cpplint` warning is for real.

---

Review status: all files reviewed at latest revision, 2 unresolved discussions, some commit checks failed.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/3105#-:-KPJsxrGUy9a4G486YkE:b3nhhd)*
<!-- Sent from Reviewable.io -->

naveenoid commented:
Review status: 8 of 10 files reviewed at latest revision, 2 unresolved discussions.

---

*[drake/examples/kuka_iiwa_arm/test/kuka_iiwa_gravity_compensation_test.cc, line 53 \[r5\]](https://reviewable.io:443/reviews/robotlocomotion/drake/3105#-KPIs17Mu3jzxAcx6Qgi:-KPMMwHwZPJFwqZT9vEl:b-896fja) ([raw file](https://github.com/robotlocomotion/drake/blob/283c9a1b03c8805e3f06c71ff48aaa7d93996084/drake/examples/kuka_iiwa_arm/test/kuka_iiwa_gravity_compensation_test.cc#L53)):*
<details><summary><i>Previously, liangfok (Chien-Liang Fok) wrote…</i></summary>
> Looks good the way you have it :smile: .
</details>
Done!

---

*[drake/systems/gravity_compensated_pd_position_control_system.h, line 22 \[r1\]](https://reviewable.io:443/reviews/robotlocomotion/drake/3105#-KOph5vIs6_ZVJG0Z-jI:-KPMMSAcLnVbWGr59tGV:b-896fix) ([raw file](https://github.com/robotlocomotion/drake/blob/4c45ef209290d3cdf017ebc7105da7a65a7061ec/drake/systems/gravity_compensated_pd_position_control_system.h#L22)):*
<details><summary><i>Previously, liangfok (Chien-Liang Fok) wrote…</i></summary>
> We should be able to anticipate the URL of a file once it's committed to master. Something like:
> 
> ```
> <img src="https://github.com/RobotLocomotion/drake/tree/master/drake/doc/images/foo.png" />
> ```
> 
> According to [this page](https://www.stack.nl/~dimitri/doxygen/manual/htmlcmds.html) the above can be directly included within a Doxygen comment block.
</details>
Done.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/3105)*
<!-- Sent from Reviewable.io -->

naveenoid commented:
BTW cpplint stuff is fixed.

---

Review status: 8 of 10 files reviewed at latest revision, 2 unresolved discussions.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/3105#-:-KPNs8ymSk692eAt7mkT:b-szbzwg)*
<!-- Sent from Reviewable.io -->

sammy-tri commented:
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/>

---

Reviewed 3 of 6 files at r6, 2 of 2 files at r7.
Review status: all files reviewed at latest revision, 2 unresolved discussions.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/3105#-:-KPNsKcpfEuROHo0Mpxj:bnfp4nl)*
<!-- Sent from Reviewable.io -->

liangfok commented:
Just one BTW nit. Otherwise, <img class="emoji" title=":lgtm_strong:" alt=":lgtm_strong:" align="absmiddle" src="https://reviewable.io/lgtm_strong.png" height="20" width="61"/> .

---

Review status: all files reviewed at latest revision, 1 unresolved discussion, all commit checks successful.

---

*[drake/systems/gravity_compensated_pd_position_control_system.h, line 85 \[r7\]](https://reviewable.io:443/reviews/robotlocomotion/drake/3105#-KPNtgfmMLO8ddI1mpAH:-KPNtgfmMLO8ddI1mpAI:bhgpwol) ([raw file](https://github.com/robotlocomotion/drake/blob/e4af44b330a447bff16d14acf6ff43106d5d3e6c/drake/systems/gravity_compensated_pd_position_control_system.h#L85)):*
> ```objc
>     vd.setZero();
> 
>     // The generalised gravity effort is computed by calling inverse dynamics
> ```

BTW: nit: UK spelling of "generalized".

http://grammarist.com/spelling/generalize-vs-generalise/

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/3105#-:-KPNu-e6LB7P3I4xMxQJ:bn4b8wv)*
<!-- Sent from Reviewable.io -->

liangfok commented:
Reviewed 1 of 6 files at r6, 2 of 2 files at r7.
Review status: all files reviewed at latest revision, 1 unresolved discussion.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/3105)*
<!-- Sent from Reviewable.io -->

naveenoid commented:
Review status: all files reviewed at latest revision, 1 unresolved discussion, all commit checks successful.

---

*[drake/systems/gravity_compensated_pd_position_control_system.h, line 85 \[r7\]](https://reviewable.io:443/reviews/robotlocomotion/drake/3105#-KPNtgfmMLO8ddI1mpAH:-KPNv3jPFNRjf_lgwaiV:b-896fix) ([raw file](https://github.com/robotlocomotion/drake/blob/e4af44b330a447bff16d14acf6ff43106d5d3e6c/drake/systems/gravity_compensated_pd_position_control_system.h#L85)):*
<details><summary><i>Previously, liangfok (Chien-Liang Fok) wrote…</i></summary>
> BTW: nit: UK spelling of "generalized".
> 
> http://grammarist.com/spelling/generalize-vs-generalise/
</details>
Done.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/3105)*
<!-- Sent from Reviewable.io -->

liangfok commented:
Reviewed 1 of 6 files at r6, 1 of 1 files at r8.
Review status: all files reviewed at latest revision, all discussions resolved, some commit checks pending.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/3105)*
<!-- Sent from Reviewable.io -->

naveenoid commented:
Jenkins is failing with this error : 
error: downloading 'ftp://ftp.ifor.math.ethz.ch/pub/fukuda/cdd/cddlib-094h.tar.gz' failed
  status_code: 9
  status_string: "Access denied to remote resource"

Can I assume that it is spurious?

jwnimmer-tri commented:
Known issue #2771, all PRs are broken until fixed (or the ftp server starts working again).

naveenoid commented:
@drake-jenkins-bot retest this please

---

Review status: all files reviewed at latest revision, all discussions resolved, some commit checks failed.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/3105#-:-KPSjmEFzARe37Yq1d41:bzdwe9o)*
<!-- Sent from Reviewable.io -->

naveenoid commented:
Could I get one more quick check of my last commit? I had to make changes due to rebase..

---

Review status: 7 of 10 files reviewed at latest revision, all discussions resolved, some commit checks pending.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/3105#-:-KPXgG2XwbtngHOLEovX:bk1n5rb)*
<!-- Sent from Reviewable.io -->

liangfok commented:
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/>

---

Reviewed 3 of 3 files at r10.
Review status: all files reviewed at latest revision, all discussions resolved, some commit checks pending.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/3105#-:-KPXhVQSla1jzfdymaoW:bnfp4nl)*
<!-- Sent from Reviewable.io -->
